### PR TITLE
Full Windows support, decreased network bandwitch usage and labels matched by reggex

### PR DESCRIPTION
**Pull request by atgiovannini**
* Now the execution procedure differs between Unix and Win32 Operating Systems and performs each execution with the native shell.
* Each slave has its own label installation cache.
* Subversioning by hashing of any script changes between last cached and last configured.
* Each predefined setup config are checked individually.
* If the slave has installed last setup config version it will avoid following executions.
* If slave fails in some setup, will not restart all previous executions.
* Do not copy files if they were previously copied (using subversioning).
* Labels are now matched as a unique label like last versions but now, included regex support.
* A lot of code review, comments and JavaDoc updating

### CODE REVIEW

**Comment by peppelan** [comment]
@atgiovannini thanks for the contribution, very appreciated, for both the long-awaited new feature and the housekeeping.

I will review and let you know asap - I'll try to make it by the end of this week if this works for you

**Comment by atgiovannini** [comment]
Nice, I'm glad to hear that. I'll wait for merge.

File: src/main/java/org/jenkinsci/plugins/slave_setup/ComputerListenerImpl.java [hunk 1]
```diff
@@ -6,51 +6,46 @@
 import hudson.model.TaskListener;
 import hudson.remoting.Channel;
 import hudson.slaves.ComputerListener;
+// import jenkins.model.Jenkins;
```
Comment (peppelan) [inline src/main/java/org/jenkinsci/plugins/slave_setup/ComputerListenerImpl.java]: Is there a reason for keeping this?

File: src/main/java/org/jenkinsci/plugins/slave_setup/ComputerListenerImpl.java [hunk 2]
```diff
@@ -6,51 +6,46 @@
...
+     * @throws IOException IO error accessing file on disk (cache)
+     * @throws InterruptedException Pipe Broken
      */
     @Override
-    public void preOnline(Computer c, Channel channel, FilePath root, TaskListener listener) throws IOException, InterruptedException {
-        listener.getLogger().println("just before slave " + c.getName() + " gets online ...");
+    public void preOnline(Computer c, Channel channel, FilePath root, TaskListener listener)
+            throws IOException, InterruptedException {
+
+        // Componentes printer will print only debug messages.
```
Comment (peppelan) [inline src/main/java/org/jenkinsci/plugins/slave_setup/ComputerListenerImpl.java]: Maybe a typo here?

File: src/main/java/org/jenkinsci/plugins/slave_setup/SetupConfigItem.java [hunk 3]
```diff
@@ -91,7 +83,7 @@ public String getPrepareScript() {
     /**
      * Sets the prepare script code
      *
-     * @param prepareScript
+     * @param prepareScript User values
```
Comment (peppelan) [inline src/main/java/org/jenkinsci/plugins/slave_setup/SetupConfigItem.java]: not sure this change improves the reader's understanding

File: src/main/java/org/jenkinsci/plugins/slave_setup/SetupConfigItem.java [hunk 4]
```diff
@@ -100,7 +92,7 @@ public void setPrepareScript(String prepareScript) {
     /**
      * Returns the directory containing the setup relevant files and sub directories
      *
-     * @return
+     * @return User values
```
Comment (peppelan) [inline src/main/java/org/jenkinsci/plugins/slave_setup/SetupConfigItem.java]: same here

File: src/main/java/org/jenkinsci/plugins/slave_setup/SetupConfigItem.java [hunk 5]
```diff
@@ -173,6 +165,8 @@ public void setPrepareScriptExecuted(boolean prepareScriptExecuted) {
 
     /**
      * Gets the textual representation of the assigned label as it was entered by the user.
+     * 
+     * @return User values
```
Comment (peppelan) [inline src/main/java/org/jenkinsci/plugins/slave_setup/SetupConfigItem.java]: and here

File: src/test/java/org/jenkinsci/plugins/slave_setup/SetupDeployerTest.java [hunk 6]
```diff
@@ -112,73 +123,59 @@ public void testDeployToComputer() {
...
-            } catch (InterruptedException e) {
-                e.printStackTrace();
-                fail(e.getMessage());
-            }
-        }
-    }
+    // @Test
+    // public void testDeployToComputers2() {
+
+    //     // TODO: I think this test isn't necesary, we don't copy in multiple slaves
```
Comment (peppelan) [inline src/test/java/org/jenkinsci/plugins/slave_setup/SetupDeployerTest.java]: agreed, maybe better removing it

**Comment by nvasion** [comment]
Anything going on with this merge? I could really use this feature. 

Thank you.

**Comment by peppelan** [comment]
Hey, I missed the new commits fixing the CI.
Good to go!

**Pull request merged**